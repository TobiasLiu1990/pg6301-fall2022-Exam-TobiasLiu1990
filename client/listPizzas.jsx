import React, { useState } from "react";
import { useLoader } from "./lib/useLoader";
import { useNavigate } from "react-router-dom";

export function PizzaCard({ pizza: { pizza, price, ingredients, allergens } }) {
  return (
    <div>
      <h3 style={{ margin: "0", border: "0" }}>Pizza: {pizza}</h3>
      <h4 style={{ margin: "0", border: "0" }}>Price: {price}</h4>
      <h4>
        Ingredients: <IngredientsCard ingredients={ingredients} />
      </h4>
      <h4>
        Allergens: <AllergensCard allergens={allergens} />
      </h4>
    </div>
  );
}

export function IngredientsCard({ ingredients }) {
  return (
    <ul style={{ margin: "0", border: "0" }}>
      {ingredients.map((i) => (
        <li key={i}>{i}</li>
      ))}
    </ul>
  );
}

export function AllergensCard({ allergens }) {
  return (
    <ul style={{ margin: "0", border: "0" }}>
      {allergens.map((a) => (
        <li key={a}>{a}</li>
      ))}
    </ul>
  );
}

export function ListPizzas({ pizzaApi }) {
  const { loading, error, data } = useLoader(async () => {
    return await pizzaApi.listPizzas();
  });
  const navigate = useNavigate();

  if (loading) {
    return <div>Loading menu...</div>;
  }

  if (error) {
    return (
      <div
        id="error-list-pizzas-div"
        style={{ border: "solid red 1px", background: "red" }}
      >
        An error has occurred in ShowMenu(): {error.toString()}
      </div>
    );
  }

  function handleSubmitBack(e) {
    e.preventDefault();
    navigate("/");
  }

  return (
    <div>
      <form onSubmit={handleSubmitBack}>
        <button>Go back to start page</button>
      </form>

      <hr></hr>

      <h1>Our attractive menu</h1>

      <div>
        {data.map((pizza) => (
          <PizzaCard key={pizza.pizza} pizza={pizza} />
        ))}
      </div>
    </div>
  );
}
