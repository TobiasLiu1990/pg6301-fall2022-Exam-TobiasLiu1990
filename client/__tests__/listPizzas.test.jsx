import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { ListPizzas } from "../listPizzas";

const pizzas = [
  {
    pizza: "pizza 1",
    price: 100,
    ingredients: ["tomato", "cheese"],
    allergens: ["casein"],
  },
  {
    pizza: "pizza 2",
    price: 500,
    ingredients: ["tomato", "cheese", "peanuts"],
    allergens: ["casein", "nuts"],
  },
];

//Test useLoader
describe("ListPizzas component", () => {
  it("should show loading screen", () => {
    const element = document.createElement("div");
    ReactDOM.render(<ListPizzas />, element);

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show pizzas", async () => {
    const element = document.createElement("div");

    await act(async () => {
      ReactDOM.render(
        <ListPizzas
          pizzaApi={{
            listPizzas: () => new Promise((resolve) => resolve(pizzas)),
          }}
        />,
        element
      );
    });

    expect(element.querySelector("h3").innerHTML).toEqual("Pizza: " + pizzas[0].pizza);


    expect(
      Array.from(element.querySelectorAll("h3")).map((e) => e.innerHTML)
    ).toEqual(["Pizza: pizza 1", "Pizza: pizza 2"]);

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show error message", async () => {
    const element = document.createElement("div");

    await act(async () => {
      ReactDOM.render(
        <ListPizzas
          listPizzas={() => {
            throw new Error("Error, something went wrong");
          }}
        />,
        element
      );
    });

    expect(element.querySelector("#error-list-pizzas-div").innerHTML).toEqual(
      "Error, something went wrong"
    );
  });
});
