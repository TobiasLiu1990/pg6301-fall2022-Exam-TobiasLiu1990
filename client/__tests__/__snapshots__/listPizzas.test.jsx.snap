// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPizzas component should show loading screen 1`] = `"<div>Loading menu...</div>"`;

exports[`ListPizzas component should show pizzas 1`] = `"<div><form><button>Go back to start page</button></form><hr><h1>Our attractive menu</h1><div><div><h3 style="margin: 0px; border: 0px;">Pizza: pizza 1</h3><h4 style="margin: 0px; border: 0px;">Price: 100</h4><h4>Ingredients: <ul style="margin: 0px; border: 0px;"><li>tomato</li><li>cheese</li></ul></h4><h4><div>Allergens: <ul style="margin: 0px; border: 0px;"><li>casein</li></ul></div></h4></div><div><h3 style="margin: 0px; border: 0px;">Pizza: pizza 2</h3><h4 style="margin: 0px; border: 0px;">Price: 500</h4><h4>Ingredients: <ul style="margin: 0px; border: 0px;"><li>tomato</li><li>cheese</li><li>peanuts</li></ul></h4><h4><div>Allergens: <ul style="margin: 0px; border: 0px;"><li>casein</li><li>nuts</li></ul></div></h4></div></div></div>"`;
